internal/sim/api.go

```diff
@@ -0,0 +1,340 @@
+// limitations under the License.
+
+// Package sim...
+//
+// TODO(mwhittaker): Write comprehensive package documentation with examples.
+// We also probably want to put some of this documentation on the website, and
+// we might also want to write a blog.
+//
+// TODO(mwhittaker): Move things to the weavertest package.
+package sim
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ServiceWeaver/weaver/internal/reflection"
+	"github.com/ServiceWeaver/weaver/weavertest"
+)
+
+// A Generator[T] generates random values of type T.
+type Generator[T any] interface {
+	// Generate returns a randomly generated value of type T. While Generate is
+	// "random", it must be deterministic. That is, given the same instance of
+	// *rand.Rand, Generate must always return the same value.
+	Generate(*rand.Rand) T
+}
+
+// A Registrar is used to register stuff with a simulator.
+type Registrar interface {
+	// RegisterFake registers a fake implementation of a component.
+	RegisterFake(weavertest.FakeComponent)
+
+	// RegisterGenerators registers generators for a workload method, one
+	// generator per method argument. The number and type of the registered
+	// generators must match the method. For example, given the method:
+	//
+	//     Foo(context.Context, int, bool) error
+	//
+	// we must register a Generator[int] and a Generator[bool]:
+	//
+	//     var r Registrar = ...
+	//     var i Generator[int] = ...
+	//     var b Generator[bool] = ...
+	//     r.RegisterGenerators("Foo", i, b)
+	RegisterGenerators(method string, generators ...any)
+}
+
+// A Workload defines the set of operations to run as part of a simulation.
+// Every workload is defined as a named struct. To perform a simulation, a
+// simulator constructs an instance of the struct, calls the struct's Init
+// method, and then randomly calls the struct's exported methods. For example,
+// the following is a simple workload:
+//
+//	type workload struct {}
+//	func (w *workload) Init(r sim.Registrar) {...}
+//	func (w *workload) Foo(context.Context, int) error {...}
+//	func (w *workload) Bar(context.Context, bool, string) error {...}
+//	func (w *workload) baz(context.Context) error {...}
+//
+// When this workload is simulated, its Foo and Bar methods will be called with
+// random values generated by the generators registered in the Init method (see
+// [Registrar] for details). Note that unexported methods, like baz, are
+// ignored.
+//
+// Note that every exported workload method must receive a context.Context as
+// its first argument and must return a single error value. A simulation is
+// aborted when a method returns a non-nil error.
+//
+// TODO(mwhittaker): For now, the Init method is required. In the future, we
+// could make it optional and use default generators for methods.
+type Workload interface {
+	Init(Registrar) error
+}
+
+// WorkloadPointer[T] is a *T that implements the Workload interface.
```

Comment:
Can you expand on why this is needed? It seems like you need it for New(), but does it have to be public?

Reply:
Yeah, it is only needed for `New`. I think technically it doesn't have to exported, but since it is a type constraint in `New` (which is exported), I think it makes the most sense for it to also be exported? As an alternative, I could remove `WorkloadPointer` and simplify `New` to the following: 

```go
New[W Workload](*testing.T) *Simulator
```

The difference is that people would have to call `New` like `New[*myWorkload]` instead of `New[myWorkload]`. What's your preference? Happy to switch.

Reply:
I like your idea of getting rid of it, because these extra type will just confuse the user (e.g., should I be implementing the WorkloadPointer interface or Workload?)

Reply:
Sounds good. Mind if I do this in a follow up PR?